### DISCUSSION TITLE
Decimal vs float

### DISCUSSION THREAD

**Discussion by gister9000** [bmoscon/orderbook#28] [Q&A]
Float is much faster, but less precise. Depending on math operations, Decimal is between 2 and 200 times slower. There are many benchmarks on the Internet, and I've confirmed it myself.

However, float is very imprecise as demonstrated by (Python 3.12.3):
```
>>> 0.1 + 0.2 == 0.3
False
``` 

So, if one's use case requires removing or adding price level which is the result of float math operations, decimal is definitely the way to go. 

However, usually, people only want local book representation which means all numbers in the book were received as a string from the exchange which means all numbers in the book are the result of `float('string')` which will be mapped to the same value each time:
```
>>> x = {float('0.3'): 100.0} 
>>> del x[float('0.3')]  # works
```

So my conclusion is that if we don't rely on operations like this one, float is a better choice: 
```
>>> del x[0.1 + 0.2]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 0.30000000000000004
```

Am I missing something?

**Comment by gister9000** [comment]
+ if we need to rely on float math operations in a rare, one-off method, we can rely on floats if we round them. 
```
>>> x = {float('0.3'): 100.0}
>>> del x[round(0.1 + 0.2, 1)]
```
I didn't explore this much, but I'd bet that round can sometimes fail too. This should probably be a second discussion, let's focus on the initial message here.

**Comment by lijiachang** [comment]
Have you tried changing the source code from Decimal to float, and did you perform benchmark tests to see how much it improved performance?

**Reply by gister9000** [reply]
Running performance_test.py:
```
C lib OrderBook
Time: 23.318291 ms
```

Replacing Decimal with float in performance_test.py:
```
C lib OrderBook
Time: 9.735107 ms
```

**Reply by lijiachang** [reply]
I just reviewed my code,   
and in fact, I am using float instead of Decimal because I don't need such high precision. I will eventually round the data I want.

**Reply by lijiachang** [reply]
In my system, numbers are propagated as float type across various components,   
and rounding is done at the final stage. I maintain a relationship between symbols and their precision（the relationship is from exchange api, update periodic).

**Reply by gister9000** [reply]
I see you do spot-futures arbitrage. I run prop trading firm and we do perpetuals-spot arb. Do you want to connect and discuss latency optimizations? feel free to send me an email to `filip[@]numbagoup.com`

**Reply by bmoscon** [reply]
@gister9000 @lijiachang - happy to take any optimizations. For float vs decimal, this presumably could be defaulted to as-is and allow for the option to do it as floats?

**Reply by gister9000** [reply]
@bmoscon Your library supports both float and decimal. 

The main point of my discussion it to see if someone will point out something wrong with my conclusion:
IF I only make floats from strings, I should get the best of both worlds: speed of floats without the inaccuracy issues of float. It would seem to be as accurate as decimal.

**Reply by lijiachang** [reply]
> I see you do spot-futures arbitrage. I run prop trading firm and we do perpetuals-spot arb. Do you want to connect and discuss latency optimizations? feel free to send me an email to `filip[@]numbagoup.com`

i have sent a email for you
